06-classification_continued.Rmd

```diff
@@ -559,16 +657,378 @@ The overall workflow for performing K-nearest neighbour classification using `ti
+
+## get accuracy if we always just guess the most frequent label
+#base_acc <- cancer_irrelevant %>%
+#                group_by(Class) %>%
+#                summarize(n = n()) %>%
+#                mutate(frac = n/sum(n)) %>%
+#                summarize(mx = max(frac)) %>%
+#                select(mx)
+#base_acc <- base_acc$mx %>% unlist()
+
+# plot
+res <- tibble(ks = ks, accs = accs, fixedaccs = fixedaccs, nghbrs = nghbrs)
+#res <- res %>% mutate(base_acc = base_acc)
+#plt_irrelevant_accuracies <- res %>%
+#  ggplot() +
+#  geom_line(mapping = aes(x=ks, y=accs, linetype="Tuned K-NN")) +
+#  geom_hline(data=res, mapping=aes(yintercept=base_acc, linetype="Always Predict Benign")) +
+#  labs(x = "Number of Irrelevant Predictors", y = "Model Accuracy Estimate") + 
+#  scale_linetype_manual(name="Method", values = c("dashed", "solid"))
+
+plt_irrelevant_accuracies <- ggplot(res) +
+              geom_line(mapping = aes(x=ks, y=accs)) +
+              labs(x = "Number of Irrelevant Predictors", y = "Model Accuracy Estimate") 
+
+plt_irrelevant_accuracies
+```
+</center>
+
+Although the accuracy decreases as expected, one surprising thing about 
+Figure \@ref(fig:06-performance-irrelevant-features) is that it shows that the method
+still outperforms the baseline majority classifier (with about `r round(cancer_propn_1[1,1], 0)`% accuracy) 
+even with 40 irrelevant variables.
+How could that be? Figure \@ref(fig:06-neighbours-irrelevant-features) provides the answer:
+the tuning procedure for the K-nearest neighbours classifier combats the extra randomness from the irrelevant variables 
+by increasing the number of neighbours. Of course, because of all the extra noise in the data from the irrelevant
+variables, the number of neighbours does not increase smoothly; but the general trend is increasing.
+Figure \@ref(fig:06-fixed-irrelevant-features) corroborates
+this evidence; if we fix the number of neighbours to $K=3$, the accuracy falls off more quickly.
+
+```{r 06-neighbours-irrelevant-features, echo = FALSE, warning = FALSE, fig.retina = 2, out.width = "600", fig.cap = "Tuned number of neighbours for varying number of irrelevant predictors"}
+plt_irrelevant_nghbrs <- ggplot(res) +
+              geom_line(mapping = aes(x=ks, y=nghbrs)) +
+              labs(x = "Number of Irrelevant Predictors", y = "Number of neighbours")
+plt_irrelevant_nghbrs
+```
+
+```{r 06-fixed-irrelevant-features, echo = FALSE, warning = FALSE, fig.retina = 2, out.width = "600", fig.cap = "Accuracy versus number of irrelevant predictors for tuned and untuned number of neighbours"}
+res_tmp <- res %>% pivot_longer(cols=c("accs", "fixedaccs"), names_to="Type", values_to="accuracy")
+plt_irrelevant_nghbrs <- ggplot(res_tmp) +
+              geom_line(mapping = aes(x=ks, y=accuracy, color=Type)) +
+              labs(x = "Number of Irrelevant Predictors", y = "Accuracy") + 
+              scale_color_discrete(labels= c("Tuned K", "K = 3"))
+plt_irrelevant_nghbrs
+```
+
```

Comment:
Are we getting into too many details in this section? Not sure, but just want to give pause and think and answer this question. The main point we want to get across is that extra predictors can sometimes negatively affect your predictions for the reasons you state above. The issue about outperforming the baseline classifier, while interesting  to us and true, might add too much complexity for our audience? We might want to sacrifice a little truth here for clarity? We could just not mentions the performance of the baseline classifier, and then figs 7.10 and 7.11, and the details you describe about them. 

Again, not sure we should do that, but we should at least think it to be sure we want to get this detailed.

Reply:
this is already in an optional section -- if the reader is here they're already digging in. i'm leaving it as is